### ISSUE TITLE
Unexpected .md source formatting changes

### ISSUE THREAD

**Issue by ivanistheone** [bangle-io/bangle-io-issues#2]
I tried editing a few files (only a few characters edited) and checked the git diffs to find lots of other markup was reformatted (to an equivalent but different choice).

A. Underline-style H1 changed to #-style heading:

<img width="455" alt="Screen Shot 2021-04-17 at 1 16 51 AM" src="https://user-images.githubusercontent.com/163966/115102772-30683080-9f1b-11eb-9500-f2b977e48e93.png">

B. Spaces were added between UL LIs

<img width="350" alt="Screen Shot 2021-04-17 at 1 16 00 AM" src="https://user-images.githubusercontent.com/163966/115102777-33632100-9f1b-11eb-92cb-37e3cef6702b.png">

It would be nice if parser can preserve source and not change sections that were not edited.

**Comment by kepta** [comment]
Hey thanks for opening the issue. I will be honest to say it is going to be hard to implement this. As bangle provides a rich text editing interface it internally converts markdown to its format and then when writing it looses that information of the markdown flavor. 

I'm running out of ideas on how to cleverly implement it, but will get back when I have anything.

**Comment by ivanistheone** [comment]
Yeah I can see how this (source formatting preserving round-trip parse and writeout) would be complicated. You'd need some fancy parser, so probably not worth the complexity it will introduce.

But, perhaps there might be a simple hack to avoid the immediate UX problem: new bangle user opens an md file for viewing, doesn't make any changes, then closes the app. expected: noop ; observed: file re-written with changed formatting.

Hacky solutions idea:
 - Before writing contents of `newcontent` to file, re-read file from disk and re-parse to get `curcontent`. Diff `curconeten` and `newcontent` and if no changes, then don't write anything to file.


Unnecessarily-complex addon:
 - if there are changes, compute the text diff `changes_diff = diff(curcontent, newcontent)`, then apply to file on disk,  `patch(unparsed_text, changes_diff)`

**Comment by ivanistheone** [comment]
For the record, I don't remember every pressing save on those files that I saw changed. Maybe there is an auto-save feature? If so, an even simpler solution would be to disable autosave until first change.

**Comment by kepta** [comment]
> For the record, I don't remember every pressing save on those files that I saw changed. Maybe there is an auto-save feature? If so, an even simpler solution would be to disable autosave until first change.

This is on my radar I will post here when it is deployed.

**Comment by kepta** [comment]
I have deployed a fix (see https://github.com/bangle-io/bangle.dev/pull/173) which should respect whether a list is tight or not. I am going to work similar fix for headings and anything that has multiple representation.

**Comment by ivanistheone** [comment]
Nice! I can confirm "tight" lists are preserved now (tested on the same file as above).

In the process, I noticed some other edits based on multiple possible representations:

C. two-spaces-at-the-end-of-a-line explicit newlines replaced with backslash

<img width="845" alt="Screen Shot 2021-04-17 at 4 02 18 PM" src="https://user-images.githubusercontent.com/163966/115125524-b2496f80-9f96-11eb-811a-fdbd85b9745f.png">



D. Italics as `*` changed to italics as `_`:

<img width="525" alt="Screen Shot 2021-04-17 at 4 01 37 PM" src="https://user-images.githubusercontent.com/163966/115125534-c2614f00-9f96-11eb-90f5-fc29a2fe9abf.png">


E. Newline at end of file got disappeared 

<img width="537" alt="Screen Shot 2021-04-17 at 4 02 57 PM" src="https://user-images.githubusercontent.com/163966/115125546-d9a03c80-9f96-11eb-8a59-938e55f744a8.png">



Multiple-repr of markdown parsing might be an endless rabbit hole so beware of trying to fix all these, but if you keep fixing them I'm happy to keep keep reportin' them...
